묵찌빠 게임 [STEP 1] 무리, 혜모리

hyemory opened the pull request:
@jjpush 
안녕하세요 제이푸시! STEP1 PR 보냅니다.
한 주간 잘 부탁드립니다! 😊

---
## 고민했던 점
### 네이밍
- 처음 짝프로그래밍을 시작할때 가위바위보`RockPaperScissors` 단어를 사용했습니다.
- 코딩을 하다보니 `RockPaperScissors`가 언급되는 부분이 많아져 오히려 가독성이 좋지 않다고 생각하여 `컴퓨터 번호`와 `사용자 번호`로 네이밍을 수정하였습니다.


### 튜플을 사용한 switch문
- 가위바위보의 경우의 수가 3개 씩 9개로 고정되기 때문에 경우의 수를 각각 튜플로 나타냈습니다.
- 컴퓨터 패와 사용자가 선택한 숫자를 담은 튜플을 case와 비교하여 게임 결과를 출력하도록 구현하였습니다.

### 코드 줄이기
```swift
// 변경 전 1
guard let rockPaperScissorsNumber = Int(rockPaperScissors) else {
    print("잘못된 입력입니다. 다시 시도해주세요.")
    return choiceRockPaperScissors()
}

guard rockPaperScissorsNumber >= 0, rockPaperScissorsNumber <= 3 else {
    print("잘못된 입력입니다. 다시 시도해주세요.")
    return choiceRockPaperScissors()
}

// 변경 전 2
    let compuerNumber: Int
    computerNumber = Int.random(in: 1...3)
```

```swift
// 변경 후 1
guard let rockPaperScissorsNumber = Int(rockPaperScissors), rockPaperScissorsNumber >= 0, rockPaperScissorsNumber <= 3 else {
    print("잘못된 입력입니다. 다시 시도해주세요.")
    return choiceRockPaperScissors()
}

// 변경 후 2
    let computerNumber = Int.random(in: 1...3)
```
- 유효성 검사를 나중에 추가하다보니 guard 구문의 코드가 여러개 생겼습니다.
- 캠퍼와 코드를 살펴보다가 합칠 수 있는 부분들을 확인하여 합쳐보았습니다.

---

## 조언을 얻고싶은 점
### 네이밍
- `RockPaperScissors`처럼 긴 특정 단어를 반복해서 사용해야할 때 이름이 너무 길어지는 것 같은데 상관없을까요?
축약형은 지양해야한다고 하는데 가독성이 너무 떨어지는것같아 조언을 구하고싶습니다.

- 옵셔널 바인딩 구문에서 벗겨낼 옵셔널을 받을 상수를 만들어야 하는데, 이름 짓기가 어렵습니다.
ex. 콘솔에 입력한`userInput`, 숫자로 변환한 `convertToInt` 등으로 지었는데 직관적이지 않은 것 같아서 조언을 구하고 싶습니다.

---

parkmuri committed:
feat: 가위바위보/종료 선택함수 추가

hyemory committed:
feat: 컴퓨터의 임의의 패 만드는 함수 추가

parkmuri committed:
feat: 비교하여 결과 출력하는 함수 추가, 기존 함수 반환값 추가

hyemory committed:
feat: 게임 종료 추가, 입력 범위 유효성 추가

parkmuri committed:
docs: 헤더 수정

parkmuri committed:
chore: import 삭제

hyemory committed:
refactor: 함수, 변수 네이밍 전체적으로 수정

corpiii commented:
>## 고민했던 점 
>### 네이밍
>- 처음 짝프로그래밍을 시작할때 가위바위보`RockPaperScissors` 단어를 사용했습니다.
>- 코딩을 하다보니 `RockPaperScissors`가 언급되는 부분이 많아져 오히려 가독성이 좋지 않다고 생각하여 `컴퓨터 번호`와 `사용자 번호`로 네이밍을 수정하였습니다.

고민하신 부분에 대해 아래에 함께 코멘트 해두었습니다 😁

---

>### 튜플을 사용한 switch문
>- 가위바위보의 경우의 수가 3개 씩 9개로 고정되기 때문에 경우의 수를 각각 튜플로 나타냈습니다.
>- 컴퓨터 패와 사용자가 선택한 숫자를 담은 튜플을 case와 비교하여 게임 결과를 출력하도록 구현하였습니다.


각각 튜플로 비교해준 부분 👍👍👍👍👍👍 

>### 코드 줄이기
>```swift
>// 변경 전 1
>guard let rockPaperScissorsNumber = Int(rockPaperScissors) else {
>    print("잘못된 입력입니다. 다시 시도해주세요.")
>    return choiceRockPaperScissors()
>}
>
>guard rockPaperScissorsNumber >= 0, >rockPaperScissorsNumber <= 3 else {
>    print("잘못된 입력입니다. 다시 시도해주세요.")
>    return choiceRockPaperScissors()
>}
>
>// 변경 전 2
>    let compuerNumber: Int
>    computerNumber = Int.random(in: 1...3)
>```
>
>```swift
>// 변경 후 1
>guard let rockPaperScissorsNumber = >Int(rockPaperScissors), rockPaperScissorsNumber >= >0, rockPaperScissorsNumber <= 3 else {
>    print("잘못된 입력입니다. 다시 시도해주세요.")
>    return choiceRockPaperScissors()
>}
>
>// 변경 후 2
>    let computerNumber = Int.random(in: 1...3)
>```
>- 유효성 검사를 나중에 추가하다보니 guard 구문의 코드가 여러개 생겼습니다.
>- 캠퍼와 코드를 살펴보다가 합칠 수 있는 부분들을 확인하여 합쳐보았습니다.

합칠 수 있는 부분을 합쳐준 점 좋습니다 💯💯💯

---

>## 조언을 얻고싶은 점
>### 네이밍
>- `RockPaperScissors`처럼 긴 특정 단어를 반복해서 사용해야할 때 이름이 너무 길어지는 것 같은데 상관없을까요?
>축약형은 지양해야한다고 하는데 가독성이 너무 떨어지는것같아 조언을 구하고싶습니다.

명확한 것이 간결한 것보다 낫습니다!
말씀해주신 것처럼 축약형은 지양하는 것이 나으나 표준으로 널리 사용하는 것은 축약어로 사용하기도 합니다! (ex. url, html)
이름이 너무 길어져서 고민이 된다면 
사용해주신 Number처럼 의미가 비슷한 다른 단어를 사용하는 것도 좋습니다! 😊

다만 Number라는 것은 제가 보기에 너무 많은 뜻을 가질 수 있다고 보여집니다.
`Int`, `Double` 자료형 일 수도 있고 메뉴 번호일 수도 있고 여기서는 가위바위보를 뜻할 수도 있겠네요.

Number라고 이름을 작성하다보면 이런 애매모호해지는 부분이 단점이 될 수 있을 것 같습니다 
좀 더 명확한 다른 이름을 찾거나 따로 분리를 하는 방법이 있을 것 같습니다

---


>- 옵셔널 바인딩 구문에서 벗겨낼 옵셔널을 받을 상수를 만들어야 하는데, 이름 짓기가 어렵습니다.
>ex. 콘솔에 입력한`userInput`, 숫자로 변환한 `convertToInt` 등으로 지었는데 직관적이지 않은 것 같아서 조언을 구하고 싶습니다.
>

네이밍.. 굉장히 어렵죠..
콘솔에 입력한 `userInput`은 제가 보기엔 user의 input 값을 받는 것이라 명확해 보이는데 
`convertToInt`는 말씀해주신 것처럼 직관적이지 않아보이네요. 

또한, 동사로 시작하고 있습니다

혹시 [API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/)을 읽어보셨나요? 
시간을 들여 꼼꼼히 한번 읽어보세요! 굉장히 많은 도움이 됩니다 😊

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
-//  Created by yagom. 
-//  Copyright © yagom academy. All rights reserved.
-// 
+//  Created by 혜모리, 무리 on 2022.12.26
```
주석 까지 신경 써 주셨네요!! 아주 좋습니다 👍

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
+//  Created by 혜모리, 무리 on 2022.12.26
+//
+
+func choiceNumber() -> Int {
```
choiceNumber 함수는 어떤기능을 하는 함수 인가요?
여러 기능을 하고 있는 것처럼 보이네요

hyemory replied:
`choiceNumber` 함수는 사용자가 가위바위보를 선택하고, 0~3 범위에 들어있는지 체크하는 함수인데
말씀하신 것처럼 두가지 기능을 하고 있어 아래와 같이 기능을 나누었습니다.
1. `choiceRockPaperScissors` : 가위바위보를 선택하는 함수
2. `checkRockPaperScissors` : 선택한 가위바위보를 0~3 범위에 들어있는지 체크하는 함수

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
+func choiceNumber() -> Int {
+    print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")
+    
+    guard let userInput = readLine(), userInput != "" else {
```
`readLine()`에서 `nil` 값이 들어오는 경우는 어떤경우 일까요? 🤔

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
+        return choiceNumber()
+    }
+
+    guard let convertToInt = Int(userInput), convertToInt >= 0, convertToInt <= 3 else {
+        print("잘못된 입력입니다. 다시 시도해주세요.")
```
이 부분이 합쳐주신 부분이군요! 👍
윗 부분과 나눠 주신 특별한 이유가 있나요??

parkmuri replied:
특별한 이유 없이 짝프로그래밍을 하면서 유저가 입력하는 값의 범위 지정을 나중에 해주어서 발생한 문제같습니다!

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
+}
 
-import Foundation
+func createComputerNumber() -> Int {
```
choiceNumber 함수의 Number는 유저의 입력으로 읽히는데
이 함수의 Number는 가위, 바위, 보를 뜻하는 것 같네요 🥲

parkmuri replied:
Number가 뜻하는 바가 애매한것같아 RockPaperScissors로 변경했습니다! 

0718e9893204bb024616d90f48043a06ff8f4305

hyemory replied:
`choiceRockPaperScissors`
`createComputerRockPaperScissors` 
함수로 좀 더 직관적으로 이름을 변경했습니다!

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
+    let compareNumbers: (Int, Int) = (createComputerNumber(), choiceNumber())
+    
+    switch compareNumbers {
+    case (1, 1), (2, 2), (3, 3):
+        print("비겼습니다!")
+        playRockPaperScissors()
+    case (1, 2), (2, 3), (3, 1):
+        print("이겼습니다!")
+    case (2, 1), (3, 2), (1, 3):
```
1, 2, 3이 각각 무엇을 뜻하나요? 명확하지 않은 것 같습니다. 😢

추가로 enum에 대해서 알고 계실까요?

parkmuri replied:
enum타입 생성하여 매직넘버 수정하였습니다!

27e52a2d09638061971e042b908ccd9112336b57

hyemory replied:
`enum`으로 `RockPaperScissors` 타입을 만들어 그 원시값을 지역 상수에 담는 것으로 표현했습니다!

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
+    switch compareNumbers {
+    case (1, 1), (2, 2), (3, 3):
+        print("비겼습니다!")
+        playRockPaperScissors()
```
재귀함수를 사용해주셨네요! 👍
재귀함수와 반복문의 차이는 무엇이 있을까요??

parkmuri replied:
반복문은 스택 메모리를 사용하지 않고, 재귀함수는 스택 메모리를 사용합니다.

반복문은 한 번에 실행되어 속도가 빠르지만 반복문을 여러번 사용하게되면 코드상에서 가독성이 좋지 않고,
재귀함수의 경우 한 번 실행될 때마다 각 스택에 쌓여 LIFO 순서대로 반환하고 빠지기 때문에 속도가 느리지만 변수 사용을 줄여 가독성이 좋다는 장점이 있습니다.


|| 반복문 | 재귀함수 |
|---|---|---|
|정의|명령을 반복적으로 실행|함수를 다시 호출하여 반복작업 수행|
|스택메모리|스택 메모리를 사용하지 않음|함수 호출 시 함수의 매개변수, 지역변수, 리턴값, 함수 종료 후 돌아가는 위치가 스택메모리에 저장|
|무한 반복 시|무한 루프는 CPU 사이클을 반복적으로 사용|무한 재귀는 스택 오버플로우 발생|
|속도|빠른 실행|느린 실행|

[참고블로그](https://yeonjewon.tistory.com/80)

jjpush commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,44 @@
 
-print("Hello, World!")
+func playRockPaperScissors() {
+    let compareNumbers: (Int, Int) = (createComputerNumber(), choiceNumber())
```
`compareNumbers`는 숫자를 비교하고 있나요? 👀
또한 동사로 시작하는 것으로 보입니다!

parkmuri committed:
refactor: 함수/변수 네이밍 수정

hyemory committed:
feat: RockPaperScissors 타입 추가 및 함수 분리

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,67 @@
-print("Hello, World!")
+func playRockPaperScissors() {
+    let computerRockPaperScissors = createComputerRockPaperScissors()
+    let userRockPaperScissors = checkRockPaperScissors(number: choiceRockPaperScissors())
```
createUserRockPaperScissors 같은 함수로 한번에 처리해줘도 좋을 것 같네요 😊

parkmuri replied:
이해가 안갔던 부분 질문드려서 답변참고하여 코드 수정했습니다! 

2200691b31e13307d71936ee75bc2621008a3534

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,67 @@
+//  Created by 혜모리, 무리 on 2022.12.26
+//
+
+enum RockPaperScissors: Int {
+    case scissor = 1
+    case rock = 2
+    case paper = 3
+}
```
enum을 잘 써주셨네요 👍
다만 main 파일에 너무 많은 함수가 있는 것 같습니다
기능에 따라 파일을 분리해보면 좋을 것 같습니다

hyemory replied:
우선 타입을 선언한 부분은 타입별로 파일을 만들어 주었습니다
컴퓨터의 패와 사용자가 선택하는 함수들을 묶어서 `ChoiceOfRockPaperScissors` 파일에 넣었고
비교, 실행하는 함수들을 묶어 `ResultOfRockPaperScissors` 에 넣었습니다.

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,67 @@
+}
 
-import Foundation
+func compareRockPaperScissors(computers: Int, users: Int) {
+    let scissor: Int = RockPaperScissors.scissor.rawValue
+    let rock: Int = RockPaperScissors.rock.rawValue
+    let paper: Int = RockPaperScissors.paper.rawValue
```
매개변수로 RockPaperScissors을 받는 것은 어떨까요? 😊

parkmuri replied:
매개변수를 `RockPaperScissors`타입으로 받게 수정하며 `func convertEnumType`을 생성 해 `computerNumber`, `userNumber`를 RockPaperScissors타입으로 반환하도록 했습니다!

corpiii commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,67 @@
+    let computerRockPaperScissors = createComputerRockPaperScissors()
+    let userRockPaperScissors = checkRockPaperScissors(number: choiceRockPaperScissors())
+    
+    compareRockPaperScissors(computers: computerRockPaperScissors, users: userRockPaperScissors)
```
가위바위보를 비교하는 중에 비기는 경우 재귀함수를 실행하네요?
실행하는 기능을 꺼내서 분리하면 좀 더 결합도가 낮은 코드가 될 것 같아요 👍

hyemory replied:
`compareRockPaperScissors`에서 비길 때 `playRockPaperScissors` 재귀함수를 실행하는 것에서
`compareRockPaperScissors`에서 비교 결과를 반환한 후 `playRockPaperScissors`에서 받아 비긴 경우 재귀함수를 실행하도록 수정했습니다.

parkmuri committed:
docs: 파일 분리

parkmuri committed:
feat: 함수 분리, 매개변수로 타입 사용 수정중

hyemory committed:
feat: GameResult 타입 추가 및 함수 수정

compareRockPaperScissors 함수 매개변수 RockPaperScissors 타입으로 수정  및 비긴 경우 playRockPaperScissors 함수를 실행하는 것에서compareRockPaperScissors 결과를 리턴 받아 playRockPaperScissors에서 비긴 경우 playRockPaperScissors를 실행하도록 수정
